beacon-chain/sync/validate_attester_slashing.go

```diff
@@ -3,96 +3,60 @@ package sync
 	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/traceutil"
 	"go.opencensus.io/trace"
 )
 
-var seenAttSlashingCacheSize = int64(1 << 10)
-
-// seenAttesterSlashings represents a cache of all the seen slashings
-var seenAttesterSlashings, _ = ristretto.NewCache(&ristretto.Config{
-	NumCounters: seenAttSlashingCacheSize,
-	MaxCost:     seenAttSlashingCacheSize,
-	BufferItems: 64,
-})
-
-func attSlashingCacheKey(slashing *ethpb.AttesterSlashing) (string, error) {
-	hash, err := hashutil.HashProto(slashing)
-	if err != nil {
-		return "", err
-	}
-	return string(hash[:]), nil
-}
-
 // Clients who receive an attester slashing on this topic MUST validate the conditions within VerifyAttesterSlashing before
 // forwarding it across the network.
-func (r *Service) validateAttesterSlashing(ctx context.Context, msg proto.Message, p p2p.Broadcaster, fromSelf bool) (bool, error) {
+func (r *Service) validateAttesterSlashing(ctx context.Context, pid peer.ID, msg *pubsub.Message) bool {
 	// The head state will be too far away to validate any slashing.
 	if r.initialSync.Syncing() {
-		return false, nil
+		return false
 	}
 
 	ctx, span := trace.StartSpan(ctx, "sync.validateAttesterSlashing")
 	defer span.End()
 
-	slashing, ok := msg.(*ethpb.AttesterSlashing)
-	if !ok {
-		return false, nil
-	}
-
-	cacheKey, err := attSlashingCacheKey(slashing)
+	m, err := r.decodePubsubMessage(msg)
 	if err != nil {
-		return false, errors.Wrapf(err, "could not hash attestation slashing")
-	}
-
-	invalidKey := invalid + cacheKey
-	if _, ok := seenAttesterSlashings.Get(invalidKey); ok {
-		return false, errors.New("previously seen invalid attester slashing received")
+		log.WithError(err).Error("Failed to decode message")
+		traceutil.AnnotateError(span, err)
+		return false
 	}
-	if _, ok := seenAttesterSlashings.Get(cacheKey); ok {
-		return false, nil
+	slashing, ok := m.(*ethpb.AttesterSlashing)
+	if !ok {
+		return false
 	}
 
 	// Retrieve head state, advance state to the epoch slot used specified in slashing message.
 	s, err := r.chain.HeadState(ctx)
 	if err != nil {
-		return false, err
+		return false
 	}
 	slashSlot := slashing.Attestation_1.Data.Target.Epoch * params.BeaconConfig().SlotsPerEpoch
 	if s.Slot < slashSlot {
 		if ctx.Err() != nil {
-			return false, errors.Wrapf(ctx.Err(),
-				"Failed to advance state to slot %d to process attester slashing", slashSlot)
+			return false
 		}
 
 		var err error
 		s, err = state.ProcessSlots(ctx, s, slashSlot)
 		if err != nil {
-			return false, errors.Wrapf(err, "Failed to advance state to slot %d", slashSlot)
+			return false
 		}
 	}
 
 	if err := blocks.VerifyAttesterSlashing(ctx, s, slashing); err != nil {
-		seenAttesterSlashings.Set(invalidKey, true /*value*/, 1 /*cost*/)
-		return false, errors.Wrap(err, "Received invalid attester slashing")
-	}
-	seenAttesterSlashings.Set(cacheKey, true /*value*/, 1 /*cost*/)
-
-	if fromSelf {
```

Comment:
does libp2p not process messages from self ?

Reply:
That's right, it won't send messages from itself to the subscription. It does, however, run validation on all messages published or subscribed. I've added a bit of a hack to immediately return true in validation if the pubsub message comes from ourselves. This bypasses the validation effectively as suggested here https://github.com/libp2p/go-libp2p-pubsub/issues/250#issuecomment-567746609